### PR TITLE
Fix func builtin in kretprobes and uretprobes

### LINKED ISSUE #68: func only matches kretprobe_trampoline on kretprobes
```
# bpftrace -e 'kretprobe:vfs_* { @[func] = count(); }'
Attaching 53 probes...
cannot attach kprobe, Invalid argument
Warning: could not attach probe kretprobe:vfs_dedupe_get_page.isra.21, skipping.
^C

@[kretprobe_trampoline]: 438
```

Would be nice if `func` could match the function on kretprobe, intsead of only ever returning `kretprobe_trampoline`. If that's not possible, it should be documented in the reference guide.

### PR DESCRIPTION
**Pull request by ajor**
For ret probes, the IP register will not be pointing to the function we want to trace. It may point to a kernel trampoline, or it may point to the caller of the traced function, as it fires after the "ret" instruction has executed.
    
Switch to using the get_func_ip BPF helper in [ku]retprobes.
    
**Disable func in [ku]retprobes** when this helper is not available, as it previously returned incorrect results.

This closes #68 and #2285.

Old behaviour:
```
# bpftrace -e 'kretprobe:vfs_read { printf("func: '%s'\n", func); exit(); }' 
Attaching 1 probe...
func: ksys_read
```

New behaviour:
```
# bpftrace -e 'kretprobe:vfs_read { printf("func: '%s'\n", func); exit(); }' 
Attaching 1 probe...
func: vfs_read
```

##### Checklist

- [ ] Language changes are updated in `man/adoc/bpftrace.adoc` and if needed in `docs/reference_guide.md`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

### CODE REVIEW

**Comment by lenticularis39** [comment]
> @lenticularis39 Do you have any idea why this test failed in CI?
> 
> ```
> 2: [  FAILED  ] builtin.func_uprobe
> 2: 	Command: ../src//bpftrace -e 'uprobe:./testprogs/uprobe_negative_retval:function1 { printf("func: \"%s\"\n", func); exit(); }'
> 2: 	Expected: func: "function1"
> 2: 	Found: Attaching 1 probe...\n__BPFTRACE_NOTIFY_PROBES_ATTACHED\nfunc: "0x400492"\n\n\n
> ```
> 
> I assume the target process must have exited before we got a chance to symbolise, but didn't your recent changes handle this case?

My fix in #2386 is not for the general case (that is why the PR doesn't close the associated issues), it is for cases where either:
- ASLR is disabled, or
- the process is running at the time of *attaching* the probe

Neither of these is true for `func_uprobe`: the process is run *after* attaching the probe, meaning that bpftrace doesn't get a chance to pre-load the correct symbol. If you change the `after` to `before` (and put a time delay into the test program), you should get the correct output most of the time (sometimes it still fails due to the bcc race condition I mentioned in https://github.com/iovisor/bpftrace/pull/2386#discussion_r1196056779).

This is also documented in the manpage: "If ASLR is enabled, user space symbolication only works when the process is running at either the time of the symbol resolution or the time of the probe attachment."

**Comment by ajor** [comment]
Doh, thank you!

File: tests/runtime/builtin [hunk 1]
```diff
@@ -96,18 +96,32 @@ EXPECT SUCCESS .*
...
-NAME func
-PROG k:vfs_read { printf("SUCCESS %s\n", func); exit(); }
-EXPECT SUCCESS .*
+NAME func_kprobe
+PROG k:vfs_read { printf("func: \"%s\"\n", func); exit(); }
+EXPECT func: "vfs_read"
+TIMEOUT 5
+AFTER ./testprogs/syscall read
+
+NAME func_kretprobe
```
Comment (viktormalik) [inline tests/runtime/builtin:105]: Shouldn't this test have a requirement for the `get_func_ip` helper?

Comment (danobi) [reply tests/runtime/builtin:105]: Yeah if you add a requirement on the feature then we don't need to disable it. It'll only run if the host supports the correct version.

**Comment by ajor** [comment]
> Also closes #2285, right?

I suppose it does fix the incorrect results described in this ticket, but I'd still like to get func working properly for uretprobes instead of just disabling it.

**Comment by ajor** [comment]
Added REQUIRES_FEATURE to runtime test and hopefully fixed the symbol resolution test issue

**Comment by danobi** [comment]
Looks like CI still failing

**Comment by ajor** [comment]
The symbolication issues must be the PID-namespacing problem (the tests run in Docker)! I was going crazy trying to understand what I was doing wrong.

Looks like we didn't have any PID-based symbolication tests that actually checked the results before this.

**Comment by viktormalik** [comment]
FWIW, kernel now supports `bpf_get_func_ip` for uretprobes:

https://git.kernel.org/bpf/bpf-next/c/a3c485a5d8d4

**Comment by jordalgo** [comment]
@ajor - do you know what other work is needed here to get this merged?

**Comment by ajor** [comment]
> @ajor - do you know what other work is needed here to get this merged?

Yes, sorry - I've got another PR in the works which adds support for PID-namespaces and is required to support these tests in CI. It just requires a little cleaning up before I submit. I'll try and get through my development backlog in January, starting next week.

File: src/bpffeature.h [hunk 2]
```diff
@@ -123,7 +123,7 @@ class BPFfeature {
   DEFINE_HELPER_TEST(probe_read_kernel_str, libbpf::BPF_PROG_TYPE_KPROBE);
   DEFINE_HELPER_TEST(ktime_get_boot_ns, libbpf::BPF_PROG_TYPE_KPROBE);
   DEFINE_HELPER_TEST(ktime_get_tai_ns, libbpf::BPF_PROG_TYPE_KPROBE);
-  DEFINE_HELPER_TEST(get_func_ip, libbpf::BPF_PROG_TYPE_TRACING);
+  DEFINE_HELPER_TEST(get_func_ip, libbpf::BPF_PROG_TYPE_KPROBE);
```
Comment (ajor) [inline src/bpffeature.h:126]: get_func_ip for kprobes was added 1 commit after it was added for tracing programs, so this is a stricter test, but realistically won't make a difference

File: tests/CMakeLists.txt [hunk 3]
```diff
@@ -14,7 +14,7 @@ if(${LLVM_VERSION_MAJOR} VERSION_EQUAL 12)
   set(CODEGEN_SRC ${CMAKE_BINARY_DIR}/tests/codegen_includes.cpp)
 else()
   set(CODEGEN_SRC "")
-  message(STATUS "Disabled codegen test for LLVM != 12")
+  message(STATUS "Disabled codegen test for LLVM != 12 (found ${LLVM_VERSION_MAJOR})")
```
Comment (ajor) [inline tests/CMakeLists.txt:17]: Drive-by fix to help me debug my environment

File: src/ast/passes/semantic_analyser.cpp [hunk 4]
```diff
@@ -346,6 +346,14 @@ void SemanticAnalyser::visit(Builtin &builtin)
         LOG(ERROR, builtin.loc, err_)
             << "The func builtin can not be used with '"
             << attach_point->provider << "' probes";
+
+      if ((type == ProbeType::kretprobe || type == ProbeType::uretprobe) &&
+          !bpftrace_.feature_->has_helper_get_func_ip()) {
+        LOG(ERROR, builtin.loc, err_)
+            << "The 'func' builtin is not available for " << type
+            << "s on kernels without the get_func_ip BPF feature. Consider "
+               "using 'probe' instead.";
```
Comment (jordalgo) [inline src/ast/passes/semantic_analyser.cpp:355]: nit: can we say either 'kprobe' or 'uprobe' depending on the context ('probe' by itself isn't a thing) ?

File: src/ast/passes/codegen_llvm.cpp [hunk 5]
```diff
@@ -327,10 +327,37 @@ void CodegenLLVM::visit(Builtin &builtin)
...
+    // the get_func_ip helper to get the instruction pointer.
+    //
+    // For [ku]retprobes, the IP register will not be pointing to the function
+    // we want to trace. It may point to a kernel trampoline, or it may point to
+    // the caller of the traced function, as it fires after the "ret"
+    // instruction has executed.
+    //
+    // The get_func_ip helper resolves these issues for us.
+    //
+    // But do not use the it for non-ret [ku]probes, as the helper will fail for
```
Comment (jordalgo) [inline src/ast/passes/codegen_llvm.cpp:341]: Curious, do you mean if folks pass an offset or if the function specified has been inlined?

**Review by jordalgo — COMMENTED** [review]
One small nit but otherwise LGTM!

File: src/ast/passes/semantic_analyser.cpp (hunk 4 above)
Comment (viktormalik) [reply src/ast/passes/semantic_analyser.cpp:355]: I think it means the `probe` builtin ;-)

**Review by viktormalik — APPROVED** [review]
LGTM!

File: src/ast/passes/semantic_analyser.cpp (hunk 4 above)
Comment (ajor) [reply src/ast/passes/semantic_analyser.cpp]: Hopefully it's clearer now!

**Review by jordalgo — APPROVED** [review]

**Pull request merged**